graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,156 @@
+
+
+def get_pagination_plan(
+    schema_info: QueryPlanningSchemaInfo, query_ast: DocumentNode, number_of_pages: int
+) -> Tuple[PaginationPlan, List[PaginationWarning]]:
+    """Make a PaginationPlan for the given query and number of desired pages if possible.
+
+    Might paginate to fewer than the desired number of pages if no good pagination plan
+    is found. This can happen when there's not enough data, or when the planner is not
+    smart enough to find a good plan. In that case it will return along with the result
+    a PaginationWarning that indicates why the pagination was not successful.
+    """
     definition_ast = get_only_query_definition(query_ast, GraphQLError)
 
+    if number_of_pages <= 0:
+        raise AssertionError(
+            u"The number of pages should be at least 1: {}".format(number_of_pages)
+        )
+    elif number_of_pages == 1:
+        return PaginationPlan(tuple()), []
+
     # Select the root node as the only vertex to paginate on.
     # TODO(bojanserafimov): Make a better pagination plan. Selecting the root is not
     #                       a good idea if:
     #                       - The root node has no pagination_key
     #                       - The root node has a unique index
     #                       - There are only a few different vertices at the root
-    pagination_node = get_only_selection_from_ast(definition_ast, GraphQLError)
+    pagination_node = get_only_selection_from_ast(definition_ast, GraphQLError).name.value
 
-    pagination_field = schema_info.pagination_keys.get(pagination_node.name.value)
+    pagination_field = schema_info.pagination_keys.get(pagination_node)
     if pagination_field is None:
-        return None
+        return PaginationPlan(tuple()), [PaginationFieldNotSpecified(pagination_node)]
 
-    if is_uuid4_type(schema_info, pagination_node.name.value, pagination_field):
-        pass
-    elif is_int_field_type(schema_info, pagination_node.name.value, pagination_field):
-        quantiles = schema_info.pagination_keys.statistics.get_field_quantiles(
-            pagination_node, pagination_field
+    if is_uuid4_type(schema_info, pagination_node, pagination_field):
+        return (
+            PaginationPlan(
+                (VertexPartition((pagination_node,), pagination_field, number_of_pages),)
+            ),
+            [],
         )
-        # We make sure there's more than enough quantiles because we don't interpolate.
-        if quantiles is None or len(quantiles) < 10 * number_of_pages:
-            return None
-    else:
-        return None
+    elif field_supports_range_reasoning(schema_info, pagination_node, pagination_field):
+        quantiles = schema_info.statistics.get_field_quantiles(pagination_node, pagination_field)
+
+        # If the ratio len(quantiles) / number_of_pages is N, then the largest page is
+        # expected to be (N + 1) / N times larger than the smallest page. So if N is 5,
```

Comment:
Is this assuming a particular distribution of data? Would you mind giving one more sentence of explanation here as to how this is derived, just enough to give the reader some intuition about why this holds?